### PR TITLE
Backfill credential management document

### PR DESCRIPTION
**Pull request by abrooksv**
## Related Issues
#944

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: designs/credentialManagement/design.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+## Specification
+
+This toolkit will introduce the following concepts:
+
+1. `ToolkitCredentialsProvider` - An abstract class that implements [AwsCredentialProvider] in the SDK. This class does not
+do and actual resolving of credentials, but instead leaves that that to concrete implementations. It instead provides
+an ID that is globally unique across all credential providers as well as defining a way to generate a display name.
+
+2. `ToolkitCredentialsProviderFactory` - Factory interface that knows how to create one or more `ToolkitCredentialsProvider`
+for a credential source. I.e. The `ProfileToolkitCredentialsProviderFactory` will create a `ProfileToolkitCredentialsProviderFactory`
```
Comment (awschristou) [inline designs/credentialManagement/design.md]: Does it create `ProfileToolkitCredentialsProviderFactory` or `ProfileToolkitCredentialsProvider` ?

File: designs/credentialManagement/design.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+AWS provides many different ways to retrieve the credentials required to make calls to AWS including but not limited to:
+credential/config files, environment variables, system properties, EC2 metadata APIs. You can also assume roles using 
+a source set of credentials or use MFA. Due to this large matrix of possibilities the toolkit strives to provide an 
+intuitive abstraction layer on top of some of these sources so that the implementation details are hidden and a 
+consistent user experience can be provided, while also enabling future expansion of new sources, including possibly by 
+third party plugins.
+
+## Specification
+
+This toolkit will introduce the following concepts:
```
Comment (awschristou) [inline designs/credentialManagement/design.md]: I'd like to see a diagram (UML, interaction, or something similar) to bring all these concepts together. The diagram could include one concrete class for each abstraction as well.

File: designs/credentialManagement/design.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+2. `ToolkitCredentialsProviderFactory` - Factory interface that knows how to create one or more `ToolkitCredentialsProvider`
+for a credential source. I.e. The `ProfileToolkitCredentialsProviderFactory` will create a `ProfileToolkitCredentialsProviderFactory`
+for each valid profile in the shared credentials file.
+
+3. `ToolkitCredentialsProviderRegistry` - This interface acts as a holder for all `ToolkitCredentialsProviderFactory`. Concrete 
+implementation uses IntelliJ extension system to load and store factories from `plugin.xml` instead of managing a hard 
+coded list. This enables the system to be extended in a dynamic manor by other plugins.
+
+4. `ToolkitCredentialsProviderManager` - This class acts as the union of all `ToolkitCredentialsProviderFactory`. Its 
+job is to be able to list all `ToolkitCredentialsProvider` and return a one when referenced by its global ID. It also
```
Comment (awschristou) [inline designs/credentialManagement/design.md]: "return a one"

File: designs/credentialManagement/design.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+has the ability to have listeners registered to it so they can listen for changes when `ToolkitCredentialsProvider` are
+added in removed such as when the shared credentials file is modified.
```
Comment (awschristou) [inline designs/credentialManagement/design.md]: "added in removed"

File: designs/credentialManagement/design.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+Example of usage:
+```xml
+    <extensions defaultExtensionNs="aws.toolkit">
+        <credentialProviderFactory implementation="software.aws.toolkits.jetbrains.core.credentials.profiles.ProfileCredentialProviderFactory"/>
+    </extensions>
+```
+
+### Credential Validation
+
+In order to validate the credentials returned by the  `ToolkitCredentialsProvider` are valid, we make a call to
```
Comment (awschristou) [inline designs/credentialManagement/design.md]: "to validate that x are valid" is redundant. Consider one of:
* "In order to verify the credentials returned by the `ToolkitCredentialsProvider` are valid,"
* "In order to validate the credentials returned by the `ToolkitCredentialsProvider`,"

File: designs/credentialManagement/design.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+```xml
+    <extensions defaultExtensionNs="aws.toolkit">
+        <credentialProviderFactory implementation="software.aws.toolkits.jetbrains.core.credentials.profiles.ProfileCredentialProviderFactory"/>
+    </extensions>
+```
+
+### Credential Validation
+
+In order to validate the credentials returned by the  `ToolkitCredentialsProvider` are valid, we make a call to 
+`sts::GetCallerInfo`.
```
Comment (awschristou) [inline designs/credentialManagement/design.md]: `getCallerIdentity`

File: designs/credentialManagement/design.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+[Credential Validation](#Credential-Validation).
+
+* If the check passes, we will mark the selected credentials active for the project window.
+* If the check fails, we will fall back to having no credentials specified and tell the user that the validation 
+failed.
+
+### Multi-Factor Authentication
+
+If the credential provider has MFA, we will need to prompt the user for their OTP. This works by blocking the 
+`resolveCredentials` call until a input dialog message prompt is filled in.
```
Comment (awschristou) [inline designs/credentialManagement/design.md]: Consider "This works by blocking the 
`resolveCredentials` call _in AwsCredentialProvider (plus link)_ ..."

File: designs/credentialManagement/design.md (hunk 1 above)
Comment (abrooksv) [reply designs/credentialManagement/design.md]: Ah ya thats not right. Will clarify and fix

File: designs/credentialManagement/design.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+a source set of credentials or use MFA. Due to this large matrix of possibilities the toolkit strives to provide an 
+intuitive abstraction layer on top of some of these sources so that the implementation details are hidden and a 
+consistent user experience can be provided, while also enabling future expansion of new sources, including possibly by 
+third party plugins.
+
+## Specification
+
+This toolkit will introduce the following concepts:
+
+1. `ToolkitCredentialsProvider` - An abstract class that implements [AwsCredentialProvider] in the SDK. This class does not
```
Comment (zhangzhx) [inline designs/credentialManagement/design.md]: The class name should be `AwsCredentialsProvider` in the plural format.

File: designs/credentialManagement/design.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+do any actual resolving of credentials, but instead leaves that that to concrete implementations. It instead provides
```
Comment (zhangzhx) [inline designs/credentialManagement/design.md]: Remove the extra `that`

File: designs/credentialManagement/design.md (hunk 8 above)
Comment (abrooksv) [reply designs/credentialManagement/design.md]: There is only 1 provider though

Comment (zhangzhx) [reply designs/credentialManagement/design.md]: I mean the `[AwsCredentialProvider] in the SDK`. The class name should be `[AwsCredentialsProvider]`

Comment (abrooksv) [reply designs/credentialManagement/design.md]: Ah

File: designs/credentialManagement/design.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+This toolkit will introduce the following concepts:
+
+1. `ToolkitCredentialsProvider` - An abstract class that implements [AwsCredentialsProvider] in the SDK. This class does not
+do any actual resolving of credentials, but instead leaves that to concrete implementations. It instead provides
+an ID that is globally unique across all credential providers as well as defining a way to generate a display name.
+
+2. `ToolkitCredentialsProviderFactory` - Factory interface that knows how to create one or more `ToolkitCredentialsProvider`
+for a credential source. A `ToolkitCredentialsProviderFactory` can create 0 or more instances of `ToolkitCredentialsProvider` 
+as long as each one is valid. Valid is defined as a profile that contains all the required
```
Comment (awschristou) [inline designs/credentialManagement/design.md]: The "Valid is defined as ..." sentences seem to apply to shared credentials, not necessarily other credentials types. Consider prefacing with "As an example, a shared credentials profile is considered valid when..."

Comment (abrooksv) [reply designs/credentialManagement/design.md]: Updated

**Pull request merged**